items/propm_allocation.py

```diff
@@ -0,0 +1,190 @@
+    the remaining subproblem. This function inserts agent i with item j to the subproblem
+    allocation.
+    """
+    for bundle in allocated_bundles:
+        for i, allocated_item in enumerate(bundle):
+            if allocated_item >= item:
+                bundle[i] = allocated_item + 1
+    allocated_bundles.insert(agent, [item])
+
+
+def divide(v: ValuationMatrix) -> List[List[int]]:
+    """"
+    In stage 1 the divider agent having index 0 partitions the goods into bundles.
+    """
+    total_value = v.verify_normalized()
+    item_order = sorted(v.objects(), key=lambda j: v[0, j])
+
+    bundles = []
+    divided_items_count = 0
+    divided_value = 0
+    for bundle_index in v.agents():
+        bundle_value = 0
+        item_index = divided_items_count
+        while item_index < v.num_of_objects and (
+                bundle_value + v[0, item_order[item_index]]) * (
+                v.num_of_agents - bundle_index) + divided_value <= total_value:
+            bundle_value += v[0, item_order[item_index]]
+            item_index += 1
+
+        bundles.append(
+            list(map(lambda t: item_order[t], range(divided_items_count, item_index))))
+        divided_items_count = item_index
+        divided_value += bundle_value
+    return bundles
+
+
+def update_decomposition(v: ValuationMatrix, agents: List[Set[int]], items: List[List[int]], bundle: List[int],
+                         candidate: int) -> Set[int]:
+    """
+    UpdateDecomposition subroutine
+    for each i (agents[i], items[i]) represents a subproblem of decomposition
+    bundle is S_t bundle
+    candidate is agent k from the paper
+    """
+    iter = len(agents) + 1
+
+    subproblem_graph = nx.DiGraph()
+    subproblem_graph.add_node(0, agents={candidate}, items=[])
+    for i in range(1, iter):
+        subproblem_graph.add_node(i, agents=agents[i - 1], items=items[i - 1])
+    subproblem_graph.add_node(iter, agents=set(), items=bundle)
+
+    subproblem_agents = nx.get_node_attributes(subproblem_graph, 'agents')
+    subproblem_items = nx.get_node_attributes(subproblem_graph, 'items')
+    for node_from in range(iter):
+        for node_to in range(1, iter + 1):
+            print(node_from, node_to, iter)
+            agent = next((a for a in subproblem_agents[node_from] if v.agent_value_for_bundle(a, subproblem_items[
+                node_to]) * v.num_of_agents >= v.verify_normalized() * len(subproblem_agents[node_to])), None)
+            if agent is not None:
+                subproblem_graph.add_edge(node_from, node_to, agent=agent)
+
+    reachable = []
+    for parent, child in nx.dfs_edges(subproblem_graph, 0):
+        nx.set_node_attributes(subproblem_graph, {child: parent}, "parent")
+        reachable.append(child)
+
+    parent = nx.get_node_attributes(subproblem_graph, "parent")
+    edge_agent = nx.get_edge_attributes(subproblem_graph, "agent")
+    if iter in reachable is not None:
+        print('kek')
```

Comment:
What is kek? Is this a debug print?
If so, it should be printed using logger.debug(...). See for example [here](https://github.com/erelsgl/fairpy/blob/master/items/round_robin.py) how logging is used.

Reply:
Yes, sorry, forgot to remove